### PR TITLE
Allow closing deleted assets

### PR DESCRIPTION
**Pull request by justicz**
I introduced a bug in #403 that makes it impossible to close an asset holding if the overall asset has been deleted. If an asset has been deleted, looking up its creator will fail. This is bad because now there is an unrecoverable minbalance, since the holding/account can never be closed.

This also uncovered a higher level issue: we want to allow people to close out assets to the creator, even if their asset holding is frozen. What does this mean for an asset that has been deleted, since that asset no longer has a creator? To resolve this, this PR assumes that it is OK to close out an asset holding if the amount is zero. I am sending an email to discuss this (also b/c it is protocol incompatible with current betanet).

I manually tested that this change allows closing out a zero asset holding for an asset that has been deleted.

Thanks to Sam Abbassi for noticing this.

### CODE REVIEW

File: data/transactions/asset.go [hunk 1]
```diff
@@ -322,44 +322,54 @@ func (ct AssetTransferTxnFields) apply(header Header, balances Balances, spec Sp
...
+		}
+
+		// Fetch the asset holding
 		sndHolding, ok := snd.Assets[ct.XferAsset]
 		if !ok {
 			return fmt.Errorf("asset %v not present in account %v", ct.XferAsset, source)
 		}
 
+		var bypassFreeze bool
+		if sndHolding.Amount == 0 {
```
Comment (zeldovich) [inline data/transactions/asset.go]: Why is this needed?  `takeOut` and `putIn` special-case a zero amount already.

File: data/transactions/asset.go [hunk 2]
```diff
@@ -322,44 +322,54 @@ func (ct AssetTransferTxnFields) apply(header Header, balances Balances, spec Sp
...
+		} else {
+			// Amount is nonzero, so asset shouldn't have been deleted.
+			// Creator should exist.
+			creatorAddr, err := balances.GetAssetCreator(ct.XferAsset)
+			if err != nil {
+				return fmt.Errorf("failed to find asset creator: %v", creatorAddr)
+			}
+
+			// Allow closing out to the asset creator even when frozen
+			if ct.AssetCloseTo == creatorAddr {
```
Comment (zeldovich) [inline data/transactions/asset.go]: Instead of special-casing this if statement, maybe look up if the destination address has the asset in its `AssetParams`?  I think that'll require no if statements, then.

Comment (justicz) [reply data/transactions/asset.go]: That seems much better, yes. And the 0 case just works because of the short circuit in `putIn` and `takeOut`

Comment (justicz) [reply data/transactions/asset.go]: That seems much better, yes. And the 0 case just works because of the short circuit in `putIn` and `takeOut`

File: ledger/accountdb.go [hunk 3]
```diff
@@ -183,6 +183,11 @@ func (qs *accountsDbQueries) lookupAssetCreator(assetIdx basics.AssetIndex) (add
 	err = db.Retry(func() error {
 		var buf []byte
 		err := qs.lookupAssetCreatorStmt.QueryRow(assetIdx).Scan(&buf)
+
+		if err == sql.ErrNoRows {
+			err = fmt.Errorf("asset %v does not exist or has been deleted", assetIdx)
```
Comment (tsachiherman) [inline ledger/accountdb.go]: nit: "asset index %d does not exist or has been deleted"

Comment (justicz) [reply ledger/accountdb.go]: updated, thanks, can you please reapprove?
